psl-core/src/main/java/org/linqs/psl/util/ModelDataCollector.java

```diff
@@ -0,0 +1,332 @@
+                    String replacement = "\'" + varConstMap.get(t.toString()) + "\')";
+                    groundedAtom = groundedAtom.replace(rightVarParen, replacement);
+                }
+            }
+        }
+        return groundedAtom;
+    }
+
+    // Parses through a formula object, and creates the needed object for modelData object consumption.
+    public static ArrayList<Object> parseFormula(Formula f, Map<String, String> varConstMap, GroundRule groundRule, boolean negation) {
+        ArrayList<Object> groundAtoms = new ArrayList<Object>();
+        if (f instanceof QueryAtom){
+            String groundedAtom = parseAtom(f, varConstMap);
+            Object decoratedFormula = decorateFormula(groundedAtom, groundRule, negation);
+            if (decoratedFormula != null) {
+                groundAtoms.add(decoratedFormula);
+            }
+        } else {
+            AbstractBranchFormula branchFormula = (AbstractBranchFormula) f;
+            for (int i = 0; i < branchFormula.length(); i++){
+                String groundedAtom = parseAtom(branchFormula.get(i), varConstMap);
+                Object decoratedFormula = decorateFormula(groundedAtom, groundRule, negation);
+                if (decoratedFormula != null) {
+                    groundAtoms.add(decoratedFormula);
+                }
+            }
+        }
+        return groundAtoms;
+    }
+
+    public static synchronized void addGroundRule(AbstractRule parentRule,
+            GroundRule groundRule, Map<Variable, Integer> variableMap,  Constant[] constantsList) {
+        if (groundRule == null) {
+            return;
+        }
+
+        // Create the variable constant map used for replacement.
+        Map<String, String> varConstMap = new HashMap<String, String>();
+        for (Map.Entry<Variable, Integer> entry : variableMap.entrySet()) {
+            varConstMap.put(entry.getKey().toString(), constantsList[entry.getValue()].rawToString());
+        }
+
+        // Captures the lhs and rhs of a ground rule in order to add to the modelData object.
+        String groundRuleString;
+        ArrayList<Object> lhs = new ArrayList<Object>();
+        ArrayList<Object> rhs = new ArrayList<Object>();
+        String operator = "";
+        if (parentRule instanceof AbstractLogicalRule) {
+            AbstractLogicalRule abstractLogicalParent = (AbstractLogicalRule) parentRule;
+            Formula formula = abstractLogicalParent.getFormula();
+            boolean negationFlag = false;
+            if (formula instanceof Implication) {
+                operator = ">>";
+                Implication implication = (Implication) formula;
+                Formula body = implication.getBody();
+                Formula head = implication.getHead();
+                if (body instanceof Negation) {
+                    Negation negation = (Negation) body;
+                    body = negation.getFormula();
+                    negationFlag = true;
+                }
```

Comment:
Newline after.
Use vertical whitespace to separate things that are logically different.